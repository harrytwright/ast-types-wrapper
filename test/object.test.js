const recast = require('recast')
const { namedTypes: n, builders: b } = require('ast-types')

const { member, computedMember, computedLiteralMember, property, object, pattern } = require('../src/object')

describe('object.member', () => {
  it('should create a member', () => {
    const dot = member('key', 'value')
    expect(n.MemberExpression.check(dot)).toBeTruthy()
  })

  it('should by variadic', () => {
    const dot = member('key', 'value', 'or', 'another', 'one')
    expect(n.MemberExpression.check(dot)).toBeTruthy()

    const code = recast.print(dot).code
    expect(code === 'key.value.or.another.one').toBeTruthy()
  })

  it('should throw on an invalid member', () => {
    assert.throw(() => member('key', 5), Error, 'Invalid identifier type: ' + 5)
  })
})

describe('object.computedMember', () => {
  it('should create a member', () => {
    const square = computedMember('key', 'value')
    expect(n.MemberExpression.check(square)).toBeTruthy()
  })

  it('should throw on invalid method', () => {
    assert.throw(() => computedMember('key', 5), Error, 'Invalid identifier type: ' + 5)
  })

  it('should by variadic', () => {
    const square = computedMember('key', 'value', 'or', 'another', 'one')
    expect(n.MemberExpression.check(square)).toBeTruthy()

    const code = recast.print(square).code
    expect(code === 'key[value][or][another][one]').toBeTruthy()
  })
})

describe('object.computedLiteralMember', () => {
  it('should create a member', () => {
    const square = computedLiteralMember('key', 'value')
    expect(n.MemberExpression.check(square)).toBeTruthy()
  })

  it('should by variadic', () => {
    const square = computedLiteralMember('key', 'value', 'or', 'another', 'one', 5)
    expect(n.MemberExpression.check(square)).toBeTruthy()

    const code = recast.print(square).code
    expect(code === 'key["value"]["or"]["another"]["one"][5]').toBeTruthy()
  })
})

describe('property', () => {
  it('should create a simple property', () => {
    const prop = property('key', 'value')
    expect(n.Property.check(prop)).toBeTruthy()
  });

  it('should create a simple property with valid code', () => {
    const prop = property('key', 'value')
    expect(n.Property.check(prop)).toBeTruthy()

    const code = recast.print(prop).code
    expect(code === 'key: "value"').toBeTruthy()
  });
});

// TODO: Maybe add more tests?? But these cover the basis
describe('object', () => {
  it('should create an object using an object', () => {
    const obj = object({
      key: 'value'
    })

    expect(n.ObjectExpression.check(obj)).toBeTruthy()
  });

  it('should create a deep object', () => {
    const obj = object({
      key: {
        key: {
          key: 'value'
        }
      }
    })

    expect(n.ObjectExpression.check(obj)).toBeTruthy()
  });

  it('should create a deep object with array', () => {
    const obj = object({
      key: {
        key: {
          key: [1, { key: 'value' }, true, ['1', '2', 3, { key: { key: 'value' } }]]
        }
      }
    })

    expect(n.ObjectExpression.check(obj)).toBeTruthy()
  });

  it(
    'should create an object using an object with custom properties',
    () => {
      const obj = object.withProperties(
        property.spread('args')
      )

      expect(n.ObjectExpression.check(obj)).toBeTruthy()
    }
  );
});

describe('pattern', () => {
  it('should create an object using an object', () => {
    const obj = pattern({
      key: 'value'
    })

    expect(n.ObjectPattern.check(obj)).toBeTruthy()
  });

  it(
    'should create an object using an object with custom properties',
    () => {
      const obj = pattern.withProperties(
        property.spread('args')
      )

      expect(n.ObjectPattern.check(obj)).toBeTruthy()
    }
  );
});
