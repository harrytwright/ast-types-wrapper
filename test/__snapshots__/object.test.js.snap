// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`object should create a deep object 1`] = `
Object {
  "comments": null,
  "loc": null,
  "properties": Array [
    Object {
      "comments": null,
      "computed": false,
      "decorators": null,
      "key": Object {
        "comments": null,
        "loc": null,
        "name": "key",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "kind": "init",
      "loc": null,
      "method": false,
      "shorthand": false,
      "type": "Property",
      "value": Object {
        "comments": null,
        "loc": null,
        "properties": Array [
          Object {
            "comments": null,
            "computed": false,
            "decorators": null,
            "key": Object {
              "comments": null,
              "loc": null,
              "name": "key",
              "optional": false,
              "type": "Identifier",
              "typeAnnotation": null,
            },
            "kind": "init",
            "loc": null,
            "method": false,
            "shorthand": false,
            "type": "Property",
            "value": Object {
              "comments": null,
              "loc": null,
              "properties": Array [
                Object {
                  "comments": null,
                  "computed": false,
                  "decorators": null,
                  "key": Object {
                    "comments": null,
                    "loc": null,
                    "name": "key",
                    "optional": false,
                    "type": "Identifier",
                    "typeAnnotation": null,
                  },
                  "kind": "init",
                  "loc": null,
                  "method": false,
                  "shorthand": false,
                  "type": "Property",
                  "value": Object {
                    "comments": null,
                    "loc": null,
                    "regex": null,
                    "type": "Literal",
                    "value": "value",
                  },
                },
              ],
              "type": "ObjectExpression",
            },
          },
        ],
        "type": "ObjectExpression",
      },
    },
  ],
  "type": "ObjectExpression",
}
`;

exports[`object should create a deep object with array 1`] = `
Object {
  "comments": null,
  "loc": null,
  "properties": Array [
    Object {
      "comments": null,
      "computed": false,
      "decorators": null,
      "key": Object {
        "comments": null,
        "loc": null,
        "name": "key",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "kind": "init",
      "loc": null,
      "method": false,
      "shorthand": false,
      "type": "Property",
      "value": Object {
        "comments": null,
        "loc": null,
        "properties": Array [
          Object {
            "comments": null,
            "computed": false,
            "decorators": null,
            "key": Object {
              "comments": null,
              "loc": null,
              "name": "key",
              "optional": false,
              "type": "Identifier",
              "typeAnnotation": null,
            },
            "kind": "init",
            "loc": null,
            "method": false,
            "shorthand": false,
            "type": "Property",
            "value": Object {
              "comments": null,
              "loc": null,
              "properties": Array [
                Object {
                  "comments": null,
                  "computed": false,
                  "decorators": null,
                  "key": Object {
                    "comments": null,
                    "loc": null,
                    "name": "key",
                    "optional": false,
                    "type": "Identifier",
                    "typeAnnotation": null,
                  },
                  "kind": "init",
                  "loc": null,
                  "method": false,
                  "shorthand": false,
                  "type": "Property",
                  "value": Object {
                    "comments": null,
                    "elements": Array [
                      Object {
                        "comments": null,
                        "loc": null,
                        "regex": null,
                        "type": "Literal",
                        "value": 1,
                      },
                      Object {
                        "comments": null,
                        "loc": null,
                        "properties": Array [
                          Object {
                            "comments": null,
                            "computed": false,
                            "decorators": null,
                            "key": Object {
                              "comments": null,
                              "loc": null,
                              "name": "key",
                              "optional": false,
                              "type": "Identifier",
                              "typeAnnotation": null,
                            },
                            "kind": "init",
                            "loc": null,
                            "method": false,
                            "shorthand": false,
                            "type": "Property",
                            "value": Object {
                              "comments": null,
                              "loc": null,
                              "regex": null,
                              "type": "Literal",
                              "value": "value",
                            },
                          },
                        ],
                        "type": "ObjectExpression",
                      },
                      Object {
                        "comments": null,
                        "loc": null,
                        "regex": null,
                        "type": "Literal",
                        "value": true,
                      },
                      Object {
                        "comments": null,
                        "elements": Array [
                          Object {
                            "comments": null,
                            "loc": null,
                            "regex": null,
                            "type": "Literal",
                            "value": "1",
                          },
                          Object {
                            "comments": null,
                            "loc": null,
                            "regex": null,
                            "type": "Literal",
                            "value": "2",
                          },
                          Object {
                            "comments": null,
                            "loc": null,
                            "regex": null,
                            "type": "Literal",
                            "value": 3,
                          },
                          Object {
                            "comments": null,
                            "loc": null,
                            "properties": Array [
                              Object {
                                "comments": null,
                                "computed": false,
                                "decorators": null,
                                "key": Object {
                                  "comments": null,
                                  "loc": null,
                                  "name": "key",
                                  "optional": false,
                                  "type": "Identifier",
                                  "typeAnnotation": null,
                                },
                                "kind": "init",
                                "loc": null,
                                "method": false,
                                "shorthand": false,
                                "type": "Property",
                                "value": Object {
                                  "comments": null,
                                  "loc": null,
                                  "properties": Array [
                                    Object {
                                      "comments": null,
                                      "computed": false,
                                      "decorators": null,
                                      "key": Object {
                                        "comments": null,
                                        "loc": null,
                                        "name": "key",
                                        "optional": false,
                                        "type": "Identifier",
                                        "typeAnnotation": null,
                                      },
                                      "kind": "init",
                                      "loc": null,
                                      "method": false,
                                      "shorthand": false,
                                      "type": "Property",
                                      "value": Object {
                                        "comments": null,
                                        "loc": null,
                                        "regex": null,
                                        "type": "Literal",
                                        "value": "value",
                                      },
                                    },
                                  ],
                                  "type": "ObjectExpression",
                                },
                              },
                            ],
                            "type": "ObjectExpression",
                          },
                        ],
                        "loc": null,
                        "type": "ArrayExpression",
                      },
                    ],
                    "loc": null,
                    "type": "ArrayExpression",
                  },
                },
              ],
              "type": "ObjectExpression",
            },
          },
        ],
        "type": "ObjectExpression",
      },
    },
  ],
  "type": "ObjectExpression",
}
`;

exports[`object should create an object using an object 1`] = `
Object {
  "comments": null,
  "loc": null,
  "properties": Array [
    Object {
      "comments": null,
      "computed": false,
      "decorators": null,
      "key": Object {
        "comments": null,
        "loc": null,
        "name": "key",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "kind": "init",
      "loc": null,
      "method": false,
      "shorthand": false,
      "type": "Property",
      "value": Object {
        "comments": null,
        "loc": null,
        "regex": null,
        "type": "Literal",
        "value": "value",
      },
    },
  ],
  "type": "ObjectExpression",
}
`;

exports[`object should create an object using an object with custom properties 1`] = `
Object {
  "comments": null,
  "loc": null,
  "properties": Array [
    Object {
      "argument": Object {
        "comments": null,
        "loc": null,
        "name": "args",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "comments": null,
      "loc": null,
      "type": "SpreadProperty",
    },
  ],
  "type": "ObjectExpression",
}
`;

exports[`object.computedLiteralMember should create a member 1`] = `
Object {
  "comments": null,
  "computed": true,
  "loc": null,
  "object": Object {
    "comments": null,
    "loc": null,
    "name": "key",
    "optional": false,
    "type": "Identifier",
    "typeAnnotation": null,
  },
  "optional": false,
  "property": Object {
    "comments": null,
    "loc": null,
    "regex": null,
    "type": "Literal",
    "value": "value",
  },
  "type": "MemberExpression",
}
`;

exports[`object.computedMember should create a member 1`] = `
Object {
  "comments": null,
  "computed": true,
  "loc": null,
  "object": Object {
    "comments": null,
    "loc": null,
    "name": "key",
    "optional": false,
    "type": "Identifier",
    "typeAnnotation": null,
  },
  "optional": false,
  "property": Object {
    "comments": null,
    "loc": null,
    "name": "value",
    "optional": false,
    "type": "Identifier",
    "typeAnnotation": null,
  },
  "type": "MemberExpression",
}
`;

exports[`object.member should create a member 1`] = `
Object {
  "comments": null,
  "computed": false,
  "loc": null,
  "object": Object {
    "comments": null,
    "loc": null,
    "name": "key",
    "optional": false,
    "type": "Identifier",
    "typeAnnotation": null,
  },
  "optional": false,
  "property": Object {
    "comments": null,
    "loc": null,
    "name": "value",
    "optional": false,
    "type": "Identifier",
    "typeAnnotation": null,
  },
  "type": "MemberExpression",
}
`;

exports[`pattern should create an object using an object 1`] = `
Object {
  "comments": null,
  "decorators": null,
  "loc": null,
  "properties": Array [
    Object {
      "comments": null,
      "computed": false,
      "decorators": null,
      "key": Object {
        "comments": null,
        "loc": null,
        "name": "key",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "kind": "init",
      "loc": null,
      "method": false,
      "shorthand": false,
      "type": "Property",
      "value": Object {
        "comments": null,
        "loc": null,
        "regex": null,
        "type": "Literal",
        "value": "value",
      },
    },
  ],
  "type": "ObjectPattern",
  "typeAnnotation": null,
}
`;

exports[`pattern should create an object using an object with custom properties 1`] = `
Object {
  "comments": null,
  "decorators": null,
  "loc": null,
  "properties": Array [
    Object {
      "argument": Object {
        "comments": null,
        "loc": null,
        "name": "args",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "comments": null,
      "loc": null,
      "type": "SpreadProperty",
    },
  ],
  "type": "ObjectPattern",
  "typeAnnotation": null,
}
`;

exports[`property should create a simple property 1`] = `
Object {
  "comments": null,
  "computed": false,
  "decorators": null,
  "key": Object {
    "comments": null,
    "loc": null,
    "name": "key",
    "optional": false,
    "type": "Identifier",
    "typeAnnotation": null,
  },
  "kind": "init",
  "loc": null,
  "method": false,
  "shorthand": false,
  "type": "Property",
  "value": Object {
    "comments": null,
    "loc": null,
    "regex": null,
    "type": "Literal",
    "value": "value",
  },
}
`;
